__tests__/services/VoiceService.test.ts

```diff
@@ -0,0 +1,219 @@
+        }),
+      });
+
+      const mockToggleVoice = jest.fn();
+      const mockTriggerSendVoiceInput = jest.fn();
+      const mockSetTextAreaValue = jest.fn();
+      const mockSetInputLength = jest.fn();
+      const mockAudioChunksRef: RefObject<BlobPart[]> = { current: [] };
+
+      const mockSettings: Settings = {
+        voice: {
+          sendAsAudio: false,
+          language: "en-US",
+          timeoutPeriod: 5000,
+          autoSendPeriod: 3000,
+        },
+      };
+
+      expect(() => {
+        startVoiceRecording(
+          mockSettings,
+          mockToggleVoice,
+          mockTriggerSendVoiceInput,
+          mockSetTextAreaValue,
+          mockSetInputLength,
+          mockAudioChunksRef,
+          mockInputRef
+        );
+      }).not.toThrow();
+    });
+  });
+
+  describe("Audio Recording", () => {
+    it("does not start MediaRecorder if microphone permissions are denied", async () => {
+      jest
+        .spyOn(navigator.mediaDevices, "getUserMedia")
+        .mockRejectedValueOnce(new Error("Permission denied"));
+
+      const mockToggleVoice = jest.fn();
+      const mockTriggerSendVoiceInput = jest.fn();
+      const mockSetTextAreaValue = jest.fn();
+      const mockSetInputLength = jest.fn();
+      const mockAudioChunksRef: RefObject<BlobPart[]> = { current: [] };
+
+      const mockSettings: Settings = {
+        voice: { sendAsAudio: true },
+      };
+
+      try {
+        await startVoiceRecording(
+          mockSettings,
+          mockToggleVoice,
+          mockTriggerSendVoiceInput,
+          mockSetTextAreaValue,
+          mockSetInputLength,
+          mockAudioChunksRef,
+          mockInputRef
+        );
+      } catch (error) {
+        if (error instanceof Error) {
+          expect(error.message).toBe("Permission denied");
+        } else {
+          throw error;
+        }
+      }
+    });
+
+    it("handles audio recording with MediaRecorder", async () => {
+      const mockToggleVoice = jest.fn();
+      const mockTriggerSendVoiceInput = jest.fn();
+      const mockSetTextAreaValue = jest.fn();
+      const mockSetInputLength = jest.fn();
+      const mockAudioChunksRef: RefObject<BlobPart[]> = { current: [] };
+
+      const mockSettings: Settings = {
+        voice: {
+          sendAsAudio: true,
+        },
+      };
+
+      navigator.mediaDevices.getUserMedia = jest
+        .fn()
+        .mockResolvedValueOnce(new MediaStream());
+
+      await startVoiceRecording(
+        mockSettings,
+        mockToggleVoice,
+        mockTriggerSendVoiceInput,
+        mockSetTextAreaValue,
+        mockSetInputLength,
+        mockAudioChunksRef,
+        mockInputRef
+      );
+
+      expect(navigator.mediaDevices.getUserMedia).toHaveBeenCalledWith({
+        audio: true,
+      });
+    });
+  });
+
+  it("stops voice recording without errors", () => {
+    stopVoiceRecording();
+    expect(true).toBe(true); // Dummy check
```

Comment:
The dummy check doesn't validate anything here, a more meaningful check would be on the state of `voiceToggledOn`.